src/thin_arc.rs

```diff
@@ -107,7 +107,11 @@ impl<H, T> ThinArc<H, T> {
 
         // Expose the transient Arc to the callback, which may clone it if it wants
         // and forward the result to the user
-        f(&mut transient)
+        let ret = f(&mut transient);
+        // It is possible for the user to replace the Arc entirely here. If so, we need to update the ThinArc as well
+        // Safety: We're still in the realm of Protected types so this cast is safe
+        self.ptr = transient.p.cast();
```

Comment:
this writing back needs to move to some guard object's `Drop` for panic safety

Reply:
Hmm, I don't see an easy way of doing that: `f` borrows `transient` mutably so I can't have another handle to it active at the same time. Not quite sure how to get around that without introducing some Refcelly business.

Reply:
oh nvm figured it out. we have the guard object own it. easy